lib/services/local/users.go

```diff
@@ -800,14 +807,53 @@ func (s *IdentityService) UpsertPassword(user string, password []byte) error {
 		return trace.Wrap(err)
 	}
 
-	err = s.UpsertPasswordHash(user, hash)
-	if err != nil {
+	if err := s.upsertPasswordHash(user, hash); err != nil {
 		return trace.Wrap(err)
 	}
 
+	_, err = s.UpdateAndSwapUser(
+		context.TODO(),
+		user,
+		false, /*withSecrets*/
+		func(u types.User) (bool, error) {
+			u.SetPasswordState(types.PasswordState_PASSWORD_STATE_SET)
+			return true, nil
+		})
+	if err != nil {
+		// Don't let the password state flag change fail the entire operation.
+		s.log.WithError(err).Info("Failed to set password state")
+	}
+
 	return nil
 }
 
+// DeletePassword deletes user's password and sets the `PasswordState` status
+// flag accordingly.
+func (s *IdentityService) DeletePassword(ctx context.Context, user string) error {
+	if user == "" {
+		return trace.BadParameter("missing username")
+	}
+
+	delErr := s.Delete(ctx, backend.Key(webPrefix, usersPrefix, user, pwdPrefix))
+
+	if delErr == nil || trace.IsNotFound(delErr) {
+		_, err := s.UpdateAndSwapUser(
+			context.TODO(),
+			user,
+			false, /*withSecrets*/
+			func(u types.User) (bool, error) {
+				u.SetPasswordState(types.PasswordState_PASSWORD_STATE_UNSET)
+				return true, nil
+			})
+		if err != nil {
+			// Don't let the password state flag change fail the entire operation.
+			s.log.WithError(err).Info("Failed to set password state")
+		}
+	}
+
+	return trace.Wrap(delErr)
```

Comment:
Suggestion: return early to reduce nesting:

```suggestion
	err := s.Delete(ctx, backend.Key(webPrefix, usersPrefix, user, pwdPrefix)
	if err != nil && !trace.IsNotFound(err) {
		return trace.Wrap(err)
	}

	if _, err := s.UpdateAndSwapUser(
		context.TODO(),
		user,
		false, /*withSecrets*/
		func(u types.User) (bool, error) {
			u.SetPasswordState(types.PasswordState_PASSWORD_STATE_UNSET)
			return true, nil
		},
	); err != nil {
		// Don't let the password state flag change fail the entire operation.
		s.log.WithError(err).Info("Failed to set password state")
	}
	
	return nil
```

Reply:
I just flipped the position of some of the ifs, but the code should still be equivalent. If it's not equivalent than I made an editing mistake, but nevertheless the style can be changed. Please take a another look, but if you'd rather not apply it it's also fine, it's only a suggestion.

Reply:
Oh, of course, it can be still done with an early return, I only need to check for this error twice. :) Got it.